File: KryossApi/src/KryossApi/Data/Entities/ExternalScan.cs
namespace KryossApi.Data.Entities;

public class ExternalScan
{
    public Guid Id { get; set; }
    public Guid OrganizationId { get; set; }
    public string Target { get; set; } = null!;
    public string Status { get; set; } = "pending";
    public DateTime? StartedAt { get; set; }
    public DateTime? CompletedAt { get; set; }
    public Guid CreatedBy { get; set; }
    public DateTime CreatedAt { get; set; }

    public Organization Organization { get; set; } = null!;
    public List<ExternalScanResult> Results { get; set; } = [];
    public List<ExternalScanFinding> Findings { get; set; } = [];
}

public class ExternalScanResult
{
    public long Id { get; set; }
    public Guid ScanId { get; set; }
    public string IpAddress { get; set; } = null!;
    public int Port { get; set; }
    public string Protocol { get; set; } = "TCP";
    public string Status { get; set; } = null!;
    public string? Service { get; set; }
    public string? Risk { get; set; }
    public string? Banner { get; set; }
    public string? ServiceName { get; set; }
    public string? ServiceVersion { get; set; }
    public string? Detail { get; set; }

    public ExternalScan Scan { get; set; } = null!;
}

public class ExternalScanFinding
{
    public long Id { get; set; }
    public Guid ScanId { get; set; }
    public long? ScanResultId { get; set; }
    public string Severity { get; set; } = null!;
    public string Title { get; set; } = null!;
    public string? Description { get; set; }
    public string? Remediation { get; set; }
    public int? Port { get; set; }
    public string? PublicIp { get; set; }
    public string? Category { get; set; }

    public ExternalScan Scan { get; set; } = null!;
    public ExternalScanResult? ScanResult { get; set; }
}

File: KryossPortal/src/api/externalScan.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import { apiFetch } from './client';

// ── Types ──

export interface ExternalScanResultItem {
  ipAddress: string;
  port: number;
  protocol: string;
  status: string;
  service: string | null;
  risk: string | null;
  banner: string | null;
  serviceName: string | null;
  serviceVersion: string | null;
}

export interface ExternalScanFindingItem {
  severity: string;
  title: string;
  description: string | null;
  remediation: string | null;
  port: number | null;
  category: string | null;
}

export interface DiscoveredTarget {
  value: string;
  source: string;
  label: string | null;
}

export interface AutoScanResult {
  scanned: number;
  scanIds: Array<{
    scanId: string | null;
    target: string;
    source: string;
    error?: string;
  }>;
}

export interface ExternalScanDetail {
  id: string;
  target: string;
  status: string;
  startedAt: string | null;
  completedAt: string | null;
  overallGrade: string | null;
  categoryScores: string | null;
  results: ExternalScanResultItem[];
  findings: ExternalScanFindingItem[];
}

export interface ScanHistoryItem {
  id: string;
  target: string;
  status: string;
  startedAt: string | null;
  completedAt: string | null;
  overallGrade: string | null;
  openPorts: number;
  criticalFindings: number;
  highFindings: number;
  totalFindings: number;
}

// ── Hooks ──

export function useExternalScanDetail(scanId: string | undefined) {
  return useQuery({
    queryKey: ['external-scan', scanId],
    queryFn: () => apiFetch<ExternalScanDetail>(`/v2/external-scan/${scanId}`),
    enabled: !!scanId,
  });
}

export function useExternalScanHistory(organizationId: string | undefined) {
  return useQuery({
    queryKey: ['external-scan-history', organizationId],
    queryFn: async () => {
      const data = await apiFetch<{ scans: ScanHistoryItem[] }>(
        `/v2/external-scan/history?organizationId=${organizationId}`,
      );
      return data.scans;
    },
    enabled: !!organizationId,
  });
}

export function useStartExternalScan() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (params: { organizationId: string; target: string }) =>
      apiFetch<{ scanId: string }>('/v2/external-scan', {
        method: 'POST',
        body: JSON.stringify(params),
      }),
    onSuccess: (_data, variables) => {
      qc.invalidateQueries({ queryKey: ['external-scan-history', variables.organizationId] });
    },
  });
}

export function useExternalScanTargets(organizationId: string | undefined) {
  return useQuery({
    queryKey: ['external-scan-targets', organizationId],
    queryFn: () =>
      apiFetch<{ targets: DiscoveredTarget[] }>(
        `/v2/external-scan/targets?organizationId=${organizationId}`,
      ),
    enabled: !!organizationId,
  });
}

export function useAutoExternalScan() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (params: { organizationId: string }) =>
      apiFetch<AutoScanResult>('/v2/external-scan/auto', {
        method: 'POST',
        body: JSON.stringify(params),
      }),
    onSuccess: (_data, variables) => {
      qc.invalidateQueries({ queryKey: ['external-scan-history', variables.organizationId] });
    },
  });
}

// ── Consent ──

export function useEnableExternalScanConsent() {
  return useMutation({
    mutationFn: (params: { organizationId: string }) =>
      apiFetch(`/v2/organizations/${params.organizationId}/external-scan`, {
        method: 'PATCH',
        body: JSON.stringify({ enabled: true }),
      }),
  });
}
